docs/specification.md

```diff
@@ -0,0 +1,226 @@
+# Type Shape Derivation Specification
+
+This document provides a comprehensive specification for how .NET types are mapped to PolyType shapes. Both built-in shape providers (source generator and reflection provider) implement identical derivation logic that maps arbitrary .NET types into type shapes of different kinds.
+
+## Overview
+
+PolyType categorizes .NET types into eight distinct shape kinds, each represented by a specific interface:
+
+- **Object** - <xref:PolyType.Abstractions.IObjectTypeShape> for general object types with properties and constructors  
+- **Enumerable** - <xref:PolyType.Abstractions.IEnumerableTypeShape> for collection types implementing @System.Collections.Generic.IEnumerable`1
+- **Dictionary** - <xref:PolyType.Abstractions.IDictionaryTypeShape> for key-value collection types
+- **Enum** - <xref:PolyType.Abstractions.IEnumTypeShape> for enumeration types
+- **Optional** - <xref:PolyType.Abstractions.IOptionalTypeShape> for nullable value types and F# options
+- **Surrogate** - <xref:PolyType.Abstractions.ISurrogateTypeShape> for types with custom marshaling
+- **Union** - <xref:PolyType.Abstractions.IUnionTypeShape> for discriminated union types
+- **Function** - <xref:PolyType.Abstractions.IFunctionTypeShape> for delegate and F# function types
+
+## Derivation Algorithm
+
+Type shape derivation follows a priority-based algorithm that checks conditions in the following order:
+
+### 1. Surrogate Types (Highest Priority)
+
+**Condition**: A type is mapped to <xref:PolyType.Abstractions.ISurrogateTypeShape> when a custom marshaler is explicitly configured.
+
+**Implementation Details**:
+- Requires an <xref:PolyType.IMarshaler`2> implementation
+- The marshaler provides bidirectional mapping between the original type and surrogate type
+- Configured via the <xref:PolyType.GenerateShapeAttribute> attribute's `Marshaler` parameter or type extensions
```

Comment:
Include mention of TypeShapeAttribute and TypeShapeExtensionsAttribute explicitly.